src/main/java/com/google/step/YOUR_PROJECT_NAME_HERE/external/StackOverflowClient.java

```diff
@@ -0,0 +1,157 @@
+package com.google.step.YOUR_PROJECT_NAME_HERE.external;
+
+import com.google.step.YOUR_PROJECT_NAME_HERE.data.Card;
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.net.*;
+import java.util.*;
+import org.apache.http.HttpEntity;
+import org.apache.http.client.ClientProtocolException;
+import org.apache.http.client.methods.CloseableHttpResponse;
+import org.apache.http.client.methods.HttpGet;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.HttpClients;
+import org.json.JSONArray;
+import org.json.JSONObject;
+import org.jsoup.Jsoup;
+import org.jsoup.nodes.Document;
+import org.jsoup.nodes.Element;
+import org.jsoup.select.Elements;
+
+public final class StackOverflowClient {
+  private static final String SEARCH_URL_TEMPLATE =
+      "https://api.stackexchange.com/2.2/questions/%s?"
+          + "order=desc&sort=activity&site=stackoverflow";
+  private static final String QUESTION_URL_TEMPLATE =
+      "https://api.stackexchange.com/2.2/questions/%s/answers?"
+          + "order=desc&sort=votes&site=stackoverflow";
+  private static final String ANSWER_URL_TEMPLATE =
+      "https://api.stackexchange.com/2.2/answers/%s?order"
+          + "=desc&sort=activity&site=stackoverflow&filter=!9_bDE(fI5";
+
+  public Card search(String so_url) {
+    Card card = getQuestion(so_url);
+    card = getAnswerId(card);
+    card = getAnswer(card);
+    return card;
+  }
+
+  /* Get the question id based on url from the CSE result */
+  private Card getQuestion(String so_url) {
+    Card card = new Card();
+    if (so_url == null) {
+      return card;
+    }
+    CloseableHttpClient httpClient = HttpClients.createDefault();
+    try {
+      URI uri = new URI(so_url);
+      System.out.println("path = " + uri.getPath());
+      // Parse the URL to get the question id.
+      String[] segments = uri.getPath().split("/");
+      String idStr = segments[2];
+      String url = String.format(SEARCH_URL_TEMPLATE, idStr);
+      CloseableHttpResponse response = httpClient.execute(new HttpGet(url));
+      if (response.getStatusLine().getStatusCode() != 200) {
+        return card;
+      }
+      HttpEntity entity = response.getEntity();
+      if (entity == null) {
+        return card;
+      }
+      BufferedReader reader = new BufferedReader(new InputStreamReader(entity.getContent()));
+      StringBuilder responseBody = new StringBuilder();
+      String line = null;
+      while ((line = reader.readLine()) != null) {
+        responseBody.append(line);
+      }
+      JSONObject res = new JSONObject(responseBody.toString());
+      String title = res.getJSONArray("items").getJSONObject(0).get("title").toString();
+      card.setLink(uri.toString());
+      // Store the id of the question in order to get the code body of the answer.
+      card.setCode(idStr);
+      card.setTitle(title);
```

Comment:
Is there a way to return this information distinctly, rather than setting it here directly as the fields of the card such that the card is only created when all pieces of information exist? i.e. have different helper functions, each responsible for returning just the code, title, desc, link etc?